faunadb/stream_subscription.go

```diff
@@ -0,0 +1,89 @@
+package faunadb
+
+import (
+	"sync"
+)
+
+func newSubscription(client *FaunaClient, query Expr, config ...StreamConfig) StreamSubscription {
+	sub := StreamSubscription{
+		query,
+		streamConfig{
+			[]StreamField{},
+		},
+		client,
+		streamConnectionStatus{status: StreamConnIdle},
+		make(chan StreamEvent),
+		make(chan bool),
+		make(chan bool),
+	}
+	for _, fn := range config {
+		fn(&sub)
+	}
+	return sub
+}
+
+type streamConnectionStatus struct {
+	mu     sync.Mutex
+	status StreamConnectionStatus
+}
+
+func (s *streamConnectionStatus) Set(status StreamConnectionStatus) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.status = status
+}
+
+func (s *streamConnectionStatus) Get() StreamConnectionStatus {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	return s.status
+}
+
+// StreamSubscription dispatches events received to the registered listener functions.
+// New subscriptions must be constructed via the FaunaClient stream method.
+type StreamSubscription struct {
+	query          Expr
+	config         streamConfig
+	client         *FaunaClient
+	status         streamConnectionStatus
+	eventsMessages chan StreamEvent
+	closingMessage chan bool
+	getNext        chan bool
+}
+
+// Query returns the query used to initiate the stream
+func (sub *StreamSubscription) Query() Expr {
+	return sub.query
+}
+
+// Status returns the current stream status
+func (sub *StreamSubscription) Status() StreamConnectionStatus {
+	return sub.status.Get()
+}
+
+// Start initiates the stream subscription.
+func (sub *StreamSubscription) Start() error {
+	if sub.status.Get() != StreamConnActive {
+		go func() {
+			sub.getNext <- true
+		}()
+	}
+
+	return sub.client.startStream(sub)
+}
+
+func (sub *StreamSubscription) Close() {
+	go func() {
+		sub.closingMessage <- true
+	}()
+}
+
+func (sub *StreamSubscription) EventsMessages() chan StreamEvent {
+	return sub.eventsMessages
+}
+
+func (sub *StreamSubscription) Request() {
+	go func() {
+		sub.getNext <- true
+	}()
+}
```

Comment:
Why do you need to actively request the next message? I believe you can rely on channels for reactiveness.

Reply:
"I believe you can rely on channels for reactiveness." - not sure I understand the way to implement this.
Since we are using channels for "tracking" closing event we also need a way to perform request if there is no closing event in the channel. Also the same behavior is in jvm driver.

Reply:
This sounds to me that there is something wrong with our model here. Let me restate my understanding:

We have a coroutine in a closed loop reading messages from the network and pushing then to a go channel. Then we have an external command to close the stream which should also kill the coroutine. We've introduced a separate channel to track signal that the subscription is closing which forced us to select between channels in the closed loop. Since reading from the network is a blocking operation, we were then forced to introduce a separate channel managed to the user to signal its code is ready to process the next event.

If that's correct, I believe the usage of `select` is the issue here. We can move closing resources to a separate coroutine that gets fired on `sub.Close()`. We'll have to refactor some of the state management for that matter. Please take a look at the following patch. See if it makes sense.


[patch.txt](https://github.com/fauna/faunadb-go/files/6050318/patch.txt)

Reply:
Patch is applied and works as expected. @erickpintor could you please review one more time.